server/src/main/java/com/tisza/tarock/server/player/bot/Brain.java

```diff
@@ -0,0 +1,288 @@
+package com.tisza.tarock.server.player.bot;
+
+import com.tisza.tarock.game.*;
+import com.tisza.tarock.game.announcement.*;
+import com.tisza.tarock.game.card.*;
+import com.tisza.tarock.game.card.filter.*;
+import com.tisza.tarock.server.player.bot.figure.*;
+import com.tisza.tarock.server.player.bot.figure.paskievics.*;
+import com.tisza.tarock.server.player.bot.figure.paskievics.Trull;
+import com.tisza.tarock.server.player.bot.figure.paskievics.Volat;
+
+import java.util.*;
+import java.util.stream.*;
+
+public class Brain
+{
+	private Memory memory;
+	private final Personality personality;
+	private final Set<Figure> figures = new HashSet<>();
+
+	public Brain(Personality personality)
+	{
+		this.personality = personality;
+	}
+
+	public void announce(PlayerSeat player, AnnouncementContra announcement)
+	{
+		memory.addAnnouncement(player, announcement);
+	}
+
+	public void bid(PlayerSeat player, int bid)
+	{
+		memory.getBids().put(player, bid);
+	}
+
+	public void call(PlayerSeat player, Card card)
+	{
+		memory.setCaller(player);
+		memory.setCalledCard(card);
+	}
+
+	public void playCard(PlayerSeat player, Card card, boolean myCard)
+	{
+		memory.addPlayedCard(player, card);
+
+		if (myCard)
+		{
+			memory.getMyCards().removeCard(card);
+		}
+
+		if (memory.getCardsInTrick() == 0)
+		{
+			memory.setCurrentFirstCard(card);
+			memory.setCurrentStrongestCard(card);
+		}
+		else if (card.doesBeat(memory.getCurrentStrongestCard()))
+		{
+			memory.setCurrentStrongestCard(card);
+		}
+
+		memory.updateCardsInTrick();
+	}
+
+	@SuppressWarnings("unchecked")
+	public List<Card> fold()
+	{
+		final Map<Card, Integer> cards = new HashMap<>();
+		figures.forEach(figure -> figure.fold(memory.getMyCards().filter(new SkartableCardFilter(memory.getGameType())))
+				.forEach((key, value) ->
+				{
+					cards.putIfAbsent(key, 0);
+					cards.put(key, cards.get(key) + value);
+				}));
```

Comment:
This folding logic seems a bit strange to me. Currently every announcement affects the folding even if it won't be announced.